File: tests/subtitle_integration_test.rs
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use std::path::PathBuf;

#[test]
fn test_parse_real_srt_file() {
    // Create service pointing to test-data directory
    let service = SubRipService::new(PathBuf::from("test-data"));
    let filename = "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt";

    // Test subtitle 1
    let subtitle1 = service
        .get_by_id(filename, 1)
        .expect("Failed to get subtitle 1")
        .expect("Subtitle 1 not found");

    assert_eq!(*subtitle1.index.as_ref(), 1);
    assert_eq!(subtitle1.start_time.as_ref().num_milliseconds(), 1436);
    assert_eq!(subtitle1.end_time.as_ref().num_milliseconds(), 3481);
    assert!(subtitle1.text.as_ref().contains("Previously on"));

    // Test subtitle 2
    let subtitle2 = service
        .get_by_id(filename, 2)
        .expect("Failed to get subtitle 2")
        .expect("Subtitle 2 not found");

    assert_eq!(*subtitle2.index.as_ref(), 2);
    assert!(subtitle2.text.as_ref().contains("Hello, Harry"));

    // Test subtitle 3
    let subtitle3 = service
        .get_by_id(filename, 3)
        .expect("Failed to get subtitle 3")
        .expect("Subtitle 3 not found");

    assert_eq!(*subtitle3.index.as_ref(), 3);
    assert_eq!(subtitle3.text.as_ref(), "You are a Grey.");

    // Verify we can format them back
    let formatted1 = subtitle1.to_string();
    assert!(formatted1.contains("00:00:01,436 --> 00:00:03,481"));

    println!("Successfully tested subtitle retrieval from real SRT file");
}

#[test]
fn test_parse_all_subtitles_from_file() {
    // Create service pointing to test-data directory
    let service = SubRipService::new(PathBuf::from("test-data"));
    let filename = "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt";

    // Test retrieving subtitles with various indices
    // We'll test the first 20 subtitles (assuming sequential indices)
    let mut found_count = 0;
    let mut last_found_index = 0;

    for i in 1..=100 {
        match service.get_by_id(filename, i) {
            Ok(Some(subtitle)) => {
                // Verify index matches
                assert_eq!(
                    *subtitle.index.as_ref(),
                    i,
                    "Index mismatch at subtitle {}",
                    i
                );

                // Verify duration is positive
                assert!(
                    subtitle.duration().num_milliseconds() > 0,
                    "Non-positive duration at subtitle {}",
                    i
                );

                // Verify text is not empty
                assert!(
                    !subtitle.text.as_ref().trim().is_empty(),
                    "Empty text at subtitle {}",
                    i
                );

                found_count += 1;
                last_found_index = i;
            }
            Ok(None) => {
                // Subtitle with this index doesn't exist, continue
                continue;
            }
            Err(e) => {
                panic!("Failed to retrieve subtitle {}: {}", i, e);
            }
        }
    }

    println!(
        "Successfully retrieved and validated {} subtitles (last index: {})",
        found_count, last_found_index
    );
    assert!(found_count > 10, "Should have found more than 10 subtitles");
}

#[test]
fn test_nonexistent_subtitle() {
    let service = SubRipService::new(PathBuf::from("test-data"));
    let filename = "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt";

    // Try to get a subtitle with a very high index that likely doesn't exist
    let result = service.get_by_id(filename, 99999);

    assert!(result.is_ok());
    assert!(
        result.unwrap().is_none(),
        "Should return None for non-existent subtitle"
    );
}

#[test]
fn test_subtitle_file_not_found() {
    let service = SubRipService::new(PathBuf::from("test-data"));

    let result = service.get_by_id("nonexistent.srt", 1);

    assert!(result.is_err(), "Should return error for non-existent file");
}

#[test]
fn test_parse_well_formed_srt_file() {
    // This test verifies that the parser correctly handles well-formed SRT files
    let service = SubRipService::new(PathBuf::from("test-data"));
    let filename = "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt";

    // Test a few subtitles to ensure parsing works correctly
    let subtitle1 = service
        .get_by_id(filename, 1)
        .expect("Failed to get subtitle 1")
        .expect("Subtitle 1 not found");

    assert_eq!(*subtitle1.index.as_ref(), 1);
    assert!(subtitle1.text.as_ref().contains("Previously on"));

    let subtitle100 = service
        .get_by_id(filename, 100)
        .expect("Failed to get subtitle 100")
        .expect("Subtitle 100 not found");

    assert_eq!(*subtitle100.index.as_ref(), 100);

    println!("Successfully verified parser works with well-formed SRT files");
}

#[test]
fn test_parse_russian_srt_with_blank_lines() {
    // Test parsing of Russian subtitle file with blank lines within text
    let service = SubRipService::new(PathBuf::from("test-data"));
    let filename = "russian-with-blank-lines-in-text.rus.srt";

    let result = service.get_all(filename);
    assert!(
        result.is_ok(),
        "Failed to parse {}: {:?}",
        filename,
        result.err()
    );

    let subtitles = result.unwrap();
    assert_eq!(subtitles.len(), 4, "Expected 4 subtitles in {}", filename);

    // Verify block 3 contains both lines with blank line preserved
    let sub3 = &subtitles[2]; // 0-indexed
    assert_eq!(*sub3.index.as_ref(), 3);
    assert_eq!(
        sub3.text.as_ref(),
        "Может, он не умер.\n\nЧто?",
        "Block 3 should contain both lines separated by blank line"
    );

    // Verify other subtitles parsed correctly
    let sub1 = &subtitles[0];
    assert_eq!(*sub1.index.as_ref(), 1);
    assert_eq!(sub1.text.as_ref(), "Ранее в сериале...");

    let sub2 = &subtitles[1];
    assert_eq!(*sub2.index.as_ref(), 2);
    assert_eq!(sub2.text.as_ref(), "Залезай!");

    let sub4 = &subtitles[3];
    assert_eq!(*sub4.index.as_ref(), 4);
    assert_eq!(sub4.text.as_ref(), "Мы должны вернуться и проверить.");

    println!("Successfully parsed Russian SRT file with blank lines in subtitle text");
}

File: internal/cli/cmd/subtitle_integration_test.go
package cmd

import (
	"path/filepath"
	"strings"
	"testing"

	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// subtitleTestDataService returns a SubRipService pointing at the repo's
// test-data directory, mirroring `SubRipService::new(PathBuf::from("test-data"))`
// in the Rust integration test.
func subtitleTestDataService(t *testing.T) *subtitle.SubRipService {
	t.Helper()
	return subtitle.NewSubRipService(filepath.Join(repoRoot(t), "test-data"))
}

// TestIntegrationParseRealSrtFile ports `test_parse_real_srt_file`.
func TestIntegrationParseRealSrtFile(t *testing.T) {
	service := subtitleTestDataService(t)
	filename := "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt"

	subtitle1, err := service.GetByID(filename, 1)
	if err != nil {
		t.Fatalf("Failed to get subtitle 1: %v", err)
	}
	if subtitle1 == nil {
		t.Fatal("Subtitle 1 not found")
	}
	if subtitle1.Index.Value() != 1 {
		t.Errorf("expected index 1, got %d", subtitle1.Index.Value())
	}
	if subtitle1.StartTime.Millis() != 1436 {
		t.Errorf("expected start 1436ms, got %d", subtitle1.StartTime.Millis())
	}
	if subtitle1.EndTime.Millis() != 3481 {
		t.Errorf("expected end 3481ms, got %d", subtitle1.EndTime.Millis())
	}
	if !strings.Contains(subtitle1.Text.Value(), "Previously on") {
		t.Errorf("expected text to contain 'Previously on', got %q", subtitle1.Text.Value())
	}

	subtitle2, err := service.GetByID(filename, 2)
	if err != nil {
		t.Fatalf("Failed to get subtitle 2: %v", err)
	}
	if subtitle2 == nil {
		t.Fatal("Subtitle 2 not found")
	}
	if subtitle2.Index.Value() != 2 {
		t.Errorf("expected index 2, got %d", subtitle2.Index.Value())
	}
	if !strings.Contains(subtitle2.Text.Value(), "Hello, Harry") {
		t.Errorf("expected text to contain 'Hello, Harry', got %q", subtitle2.Text.Value())
	}

	subtitle3, err := service.GetByID(filename, 3)
	if err != nil {
		t.Fatalf("Failed to get subtitle 3: %v", err)
	}
	if subtitle3 == nil {
		t.Fatal("Subtitle 3 not found")
	}
	if subtitle3.Index.Value() != 3 {
		t.Errorf("expected index 3, got %d", subtitle3.Index.Value())
	}
	if subtitle3.Text.Value() != "You are a Grey." {
		t.Errorf("expected 'You are a Grey.', got %q", subtitle3.Text.Value())
	}

	formatted1 := subtitle1.String()
	if !strings.Contains(formatted1, "00:00:01,436 --> 00:00:03,481") {
		t.Errorf("expected formatted subtitle to contain timestamps, got %q", formatted1)
	}
}

// TestIntegrationParseAllSubtitlesFromFile ports `test_parse_all_subtitles_from_file`.
func TestIntegrationParseAllSubtitlesFromFile(t *testing.T) {
	service := subtitleTestDataService(t)
	filename := "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt"

	foundCount := 0
	lastFoundIndex := uint32(0)

	for i := uint32(1); i <= 100; i++ {
		sub, err := service.GetByID(filename, i)
		if err != nil {
			t.Fatalf("Failed to retrieve subtitle %d: %v", i, err)
		}
		if sub == nil {
			continue
		}

		if sub.Index.Value() != i {
			t.Errorf("Index mismatch at subtitle %d", i)
		}
		if sub.Duration().Milliseconds() <= 0 {
			t.Errorf("Non-positive duration at subtitle %d", i)
		}
		if strings.TrimSpace(sub.Text.Value()) == "" {
			t.Errorf("Empty text at subtitle %d", i)
		}

		foundCount++
		lastFoundIndex = i
	}

	_ = lastFoundIndex
	if foundCount <= 10 {
		t.Errorf("Should have found more than 10 subtitles, found %d", foundCount)
	}
}

// TestIntegrationNonexistentSubtitle ports `test_nonexistent_subtitle`.
func TestIntegrationNonexistentSubtitle(t *testing.T) {
	service := subtitleTestDataService(t)
	filename := "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt"

	sub, err := service.GetByID(filename, 99999)
	if err != nil {
		t.Fatalf("expected ok, got error: %v", err)
	}
	if sub != nil {
		t.Error("Should return nil for non-existent subtitle")
	}
}

// TestIntegrationSubtitleFileNotFound ports `test_subtitle_file_not_found`.
func TestIntegrationSubtitleFileNotFound(t *testing.T) {
	service := subtitleTestDataService(t)

	_, err := service.GetByID("nonexistent.srt", 1)
	if err == nil {
		t.Error("Should return error for non-existent file")
	}
}

// TestIntegrationParseWellFormedSrtFile ports `test_parse_well_formed_srt_file`.
func TestIntegrationParseWellFormedSrtFile(t *testing.T) {
	service := subtitleTestDataService(t)
	filename := "Resident.Alien.S03E01.1080p.WEB-DL.RGzsRutracker.eng.srt"

	subtitle1, err := service.GetByID(filename, 1)
	if err != nil {
		t.Fatalf("Failed to get subtitle 1: %v", err)
	}
	if subtitle1 == nil {
		t.Fatal("Subtitle 1 not found")
	}
	if subtitle1.Index.Value() != 1 {
		t.Errorf("expected index 1, got %d", subtitle1.Index.Value())
	}
	if !strings.Contains(subtitle1.Text.Value(), "Previously on") {
		t.Errorf("expected text to contain 'Previously on', got %q", subtitle1.Text.Value())
	}

	subtitle100, err := service.GetByID(filename, 100)
	if err != nil {
		t.Fatalf("Failed to get subtitle 100: %v", err)
	}
	if subtitle100 == nil {
		t.Fatal("Subtitle 100 not found")
	}
	if subtitle100.Index.Value() != 100 {
		t.Errorf("expected index 100, got %d", subtitle100.Index.Value())
	}
}

// TestIntegrationParseRussianSrtWithBlankLines ports
// `test_parse_russian_srt_with_blank_lines`.
func TestIntegrationParseRussianSrtWithBlankLines(t *testing.T) {
	service := subtitleTestDataService(t)
	filename := "russian-with-blank-lines-in-text.rus.srt"

	subtitles, err := service.GetAll(filename)
	if err != nil {
		t.Fatalf("Failed to parse %s: %v", filename, err)
	}

	if len(subtitles) != 4 {
		t.Fatalf("Expected 4 subtitles in %s, got %d", filename, len(subtitles))
	}

	sub3 := &subtitles[2] // 0-indexed
	if sub3.Index.Value() != 3 {
		t.Errorf("expected index 3, got %d", sub3.Index.Value())
	}
	if sub3.Text.Value() != "Может, он не умер.\n\nЧто?" {
		t.Errorf("Block 3 should contain both lines separated by blank line, got %q", sub3.Text.Value())
	}

	sub1 := &subtitles[0]
	if sub1.Index.Value() != 1 {
		t.Errorf("expected index 1, got %d", sub1.Index.Value())
	}
	if sub1.Text.Value() != "Ранее в сериале..." {
		t.Errorf("unexpected sub1 text: %q", sub1.Text.Value())
	}

	sub2 := &subtitles[1]
	if sub2.Index.Value() != 2 {
		t.Errorf("expected index 2, got %d", sub2.Index.Value())
	}
	if sub2.Text.Value() != "Залезай!" {
		t.Errorf("unexpected sub2 text: %q", sub2.Text.Value())
	}

	sub4 := &subtitles[3]
	if sub4.Index.Value() != 4 {
		t.Errorf("expected index 4, got %d", sub4.Index.Value())
	}
	if sub4.Text.Value() != "Мы должны вернуться и проверить." {
		t.Errorf("unexpected sub4 text: %q", sub4.Text.Value())
	}
}
